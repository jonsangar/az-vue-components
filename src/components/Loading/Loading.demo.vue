<template lang="pug">
  div
    Demo(title="Tipo")
      .demo__subtitle Star
      az-loading

      .demo__subtitle Circles
      az-loading(type="circle")

      .demo__subtitle Bars
      az-loading(type="bars")

      .demo__subtitle Blobs
      az-loading(type="blobs")

      .demo__subtitle Square
      az-loading(type="square")

    Demo(title="Color")
      az-loading(color="var(--color-danger)")
      az-loading(color="var(--color-success)")

    Demo(title="Size")
      az-loading(size="10px")
      az-loading(size="20px")
      az-loading(size="30px")
      az-loading(size="40px")
</template>

<script>
export default {};
</script>

<style lang="scss" scoped>
.az-loading {
  margin: half();
}
</style>
