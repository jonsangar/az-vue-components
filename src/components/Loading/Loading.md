---
device: true
---

# Loading

### Instalación

```javascript
import { Loading } from 'didor';

Vue.use(Loading);
```

---

### Uso

#### Tipo

```pug
az-loading
az-loading(type="circle")
az-loading(type="bars")
az-loading(type="blobs")
az-loading(type="square")
```

#### Color

```pug
az-loading(color="var(--color-danger)")
az-loading(color="var(--color-success)")
```

#### Size

```pug
az-loading(size="10px")
az-loading(size="20px")
az-loading(size="30px")
az-loading(size="40px")
```

---

### API

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| type        | Tipo de loading, opciones:<br>`star` `circle` `success` `bars` `blobs` `square` | `String`  | `star` |
| color       | Determinar el color del spinner. Acepta:<br> `variales css` `hexadecimal` `nombre`           | `String` | `var(--color-brand)`   |
| size        | Determinar el tamaño del spinner. Acepta valores en:<br>`px` `rem` `em`             | `String` | `1rem`    |
