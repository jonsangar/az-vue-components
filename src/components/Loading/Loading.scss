.az-loading {
  $rootClass: &;

  display: inline-block;
  vertical-align: middle;

  /**
   * Bars loading
   */
  &--bars {
    position: relative;
    display: inline-block;
    height: 1.6em;
    text-align: center;
  }

  &__bar {
    display: inline-block;
    width: 0.3em;
    height: 100%;
    margin: 0 0.065em;
    background-color: color(loading);
    animation: loading-bars 1.2s infinite ease-in-out;

    &:nth-of-type(2) {
      animation-delay: -1.1s;
    }
    &:nth-of-type(3) {
      animation-delay: -1s;
    }
    &:nth-of-type(4) {
      animation-delay: -0.9s;
    }
    &:nth-of-type(5) {
      animation-delay: -0.8s;
    }
  }

  @keyframes loading-bars {
    0%,
    40%,
    100% {
      transform: scaleY(0.4);
    }
    20% {
      transform: scaleY(1);
    }
  }

  /**
   * Circles loading
   */
  $circlesLoadingOffset: 187;

  &--circle {
    position: relative;
    display: inline-block;
    width: 1.6em;
    height: 1.6em;
  }

  &__svg {
    width: 1.6em;
    height: 1.6em;
    -webkit-transform-origin: center center;
        -ms-transform-origin: center center;
            transform-origin: center center;
    animation: loading-circle 1.5s linear infinite;
  }

  &__path {
    stroke: color(loading);
    stroke-dasharray: $circlesLoadingOffset;
    stroke-dashoffset: 0;
    stroke-linecap: round;
    stroke-width: 6;
    transform-origin: center;
    animation: loading-circle-dash 1.5s ease-in-out infinite;
  }

  @-webkit-keyframes loading-circle{
    100%{
      -webkit-transform: rotate(360deg);
              transform: rotate(360deg);
    }
  }

  @keyframes loading-circle {

    0%{
      -webkit-transform: rotate(0deg);
              transform: rotate(0deg);
    }
    100%{
      -webkit-transform: rotate(360deg);
              transform: rotate(360deg);
    }
  }

  @keyframes loading-circle-dash {
    0% {
      stroke-dashoffset: $circlesLoadingOffset;
    }
    50% {
      stroke-dashoffset: calc($circlesLoadingOffset / 4);
      transform: rotate(135deg);
    }
    100% {
      stroke-dashoffset: $circlesLoadingOffset;
      transform: rotate(360deg);
    }
  }

  @-webkit-keyframes loading-circle-dash {
    0% {
      stroke-dashoffset: $circlesLoadingOffset;
    }
    50% {
      stroke-dashoffset: calc($circlesLoadingOffset / 4);
      transform: rotate(135deg);
    }
    100% {
      stroke-dashoffset: $circlesLoadingOffset;
      transform: rotate(360deg);
    }
  }

  /**
   * Star loading
   */
  &--star {
    position: relative;
    display: inline-block;
    width: 1.6em;
    height: 1.6em;
  }

  &__item {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;

    &::before {
      display: block;
      width: 10%;
      height: 30%;
      margin: 0 auto;
      content: '';
      background-color: color(loading);
      border-radius: 1em;
      animation: loading-star 0.9s infinite ease-in-out both;
    }

    &:nth-of-type(2) {
      transform: rotate(40deg);

      &::before {
        animation-delay: -0.8s;
      }
    }
    &:nth-of-type(3) {
      transform: rotate(80deg);

      &::before {
        animation-delay: -0.7s;
      }
    }
    &:nth-of-type(4) {
      transform: rotate(120deg);

      &::before {
        animation-delay: -0.6s;
      }
    }
    &:nth-of-type(5) {
      transform: rotate(160deg);

      &::before {
        animation-delay: -0.5s;
      }
    }
    &:nth-of-type(6) {
      transform: rotate(200deg);

      &::before {
        animation-delay: -0.4s;
      }
    }
    &:nth-of-type(7) {
      transform: rotate(240deg);

      &::before {
        animation-delay: -0.3s;
      }
    }
    &:nth-of-type(8) {
      transform: rotate(280deg);

      &::before {
        animation-delay: -0.2s;
      }
    }
    &:nth-of-type(9) {
      transform: rotate(320deg);

      &::before {
        animation-delay: -0.1s;
      }
    }
  }

  @keyframes loading-star {
    0%,
    39%,
    100% {
      opacity: 0.2;
    }
    40% {
      opacity: 1;
    }
  }

  /**
   * Blobs loading
   */
  &--blobs {
    position: relative;
    display: inline-block;
    width: 1.6em;
    height: 1.6em;
    margin: 0 0.2em;
  }

  &__blob {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 0.55em;
    height: 0.55em;
    border: 0.1em solid color(loading);
    border-radius: 50%;
    transform: translate(-50%, -50%);

    &--top {
      top: 0;
      animation: loading-blob-top 1s infinite ease-in;
    }
    &--bottom {
      top: 100%;
      animation: loading-blob-bottom 1s infinite ease-in;
    }
    &--left {
      left: 0;
      animation: loading-blob-left 1s infinite ease-in;
    }
    &--move {
      top: 0;
      background-color: color(loading);
      animation: loading-blob-move 1s infinite ease-in;
    }

    @keyframes loading-blob-bottom {
      25%,
      50%,
      75% {
        top: 50%;
        left: 100%;
      }
      100% {
        top: 0;
        left: 50%;
      }
    }

    @keyframes loading-blob-left {
      25% {
        top: 50%;
        left: 0;
      }
      50%,
      100% {
        top: 100%;
        left: 50%;
      }
    }

    @keyframes loading-blob-top {
      50% {
        top: 0;
        left: 50%;
      }
      75%,
      100% {
        top: 50%;
        left: 0;
      }
    }

    @keyframes loading-blob-move {
      0%,
      100% {
        top: 0;
        left: 50%;
      }
      25% {
        top: 50%;
        left: 100%;
      }
      50% {
        top: 100%;
        left: 50%;
      }
      75% {
        top: 50%;
        left: 0;
      }
    }
  }

  /**
   * Square loading
   */
  &--square {
    position: relative;
    top: -0.3em;
    display: inline-block;
    width: 1.6em;
    height: 1.6em;
    margin: 0 auto;
  }

  &__square__box {
    position: absolute;
    top: 0;
    left: 0;
    width: 1.6em;
    height: 1.6em;
    background: color(loading);
    border-radius: 0.2em;
    animation: loading-square-box 0.5s linear infinite;
  }

  &__square__shadow {
    position: absolute;
    top: 1.9em;
    left: 0;
    width: 1.75em;
    height: 0.25em;
    background: color(gray1);
    border-radius: 50%;
    opacity: 0.1;
    animation: loading-square-shadow 0.5s linear infinite;
  }

  @keyframes loading-square-box {
    17% {
      border-bottom-right-radius: 3px;
    }
    25% {
      transform: translateY(0.3em) rotate(22.5deg);
    }
    50% {
      border-bottom-right-radius: 1.6em;
      transform: translateY(0.8em) scale(1, 0.9) rotate(45deg);
    }
    75% {
      transform: translateY(0.3em) rotate(67.5deg);
    }
    100% {
      transform: translateY(0) rotate(90deg);
    }
  }

  @keyframes loading-square-shadow {
    50% {
      transform: scale(1.2, 1);
    }
  }
}
