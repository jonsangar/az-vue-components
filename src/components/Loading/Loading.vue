<template lang="pug">
.az-loading(:style="{'font-size': size}")
  .az-loading--bars(v-if="type === 'bars'")
    .az-loading__bar
    .az-loading__bar
    .az-loading__bar
    .az-loading__bar
    .az-loading__bar

  .az-loading--circle(v-if="type === 'circle'")
    svg.az-loading__svg(viewBox="0 0 66 66")
      circle.az-loading__path(fill="none" cx="33" cy="33" r="30")

  .az-loading--star(v-if="type === 'star'")
    .az-loading__item
    .az-loading__item
    .az-loading__item
    .az-loading__item
    .az-loading__item
    .az-loading__item
    .az-loading__item
    .az-loading__item
    .az-loading__item

  .az-loading--blobs(v-if="type === 'blobs'")
    .az-loading__blob.az-loading__blob--top
    .az-loading__blob.az-loading__blob--bottom
    .az-loading__blob.az-loading__blob--left
    .az-loading__blob.az-loading__blob--move

  .az-loading--square(v-if="type === 'square'")
    .az-loading__square__shadow
    .az-loading__square__box
</template>

<script>
export default {
  props: {
    type: {
      type: String,
      default: 'star',
      validator: value => ['bars', 'circle', 'star', 'blobs', 'square'].indexOf(value) !== -1,
    },
    color: {
      type: String,
      default: 'var(--color-brand)',
    },
    size: {
      type: String,
      default: '1rem',
    },
  },
  watch: {
    color(val) {
      this.$el.style.setProperty('--color-az-loading', val);
    },
  },
  mounted() {
    this.$el.style.setProperty('--color-az-loading', this.color);
  },
};
</script>

<style src="./Loading.scss" lang="scss"></style>
