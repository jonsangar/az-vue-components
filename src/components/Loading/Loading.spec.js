
import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import Loading from './Loading.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('Loading.vue', () => {
  const build = () => {
    const wrapper = shallowMount(Loading, {
      localVue
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente Loading con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()

  })

  it('Renderiza el loading con estilos por defecto', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toContain('az-loading');
    expect(wrapper.find('.az-loading').attributes().style).toContain('--color-az-loading: var(--color-brand)');
  });

  it('Comprueba el validator para el tipo de loading "type"', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Loading.props.type.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('bars')).toBe(true);
    expect(validator('circle')).toBe(true);
    expect(validator('star')).toBe(true);
    expect(validator('blobs')).toBe(true);
    expect(validator('square')).toBe(true);
    expect(validator('faketype')).toBe(false);
  });

  it('Comprueba que el estilo del loading es var(--color-danger) cuando se asigna dicho color', async () => {
    const { wrapper } = build()
    // Asignamos otro color para lanzar el watch
    wrapper.setProps({
      color: 'var(--color-danger)'
    })
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.az-loading').attributes().style).toContain('--color-az-loading: var(--color-danger)');
  });
});
