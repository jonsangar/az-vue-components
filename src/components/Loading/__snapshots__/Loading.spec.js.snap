// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading.vue Comprueba renderizado del componente Loading con snapshot 1`] = `
<div class="az-loading" style="font-size: 1rem; --color-az-loading: var(--color-brand);">
  <!---->
  <!---->
  <div class="az-loading--star">
    <div class="az-loading__item"></div>
    <div class="az-loading__item"></div>
    <div class="az-loading__item"></div>
    <div class="az-loading__item"></div>
    <div class="az-loading__item"></div>
    <div class="az-loading__item"></div>
    <div class="az-loading__item"></div>
    <div class="az-loading__item"></div>
    <div class="az-loading__item"></div>
  </div>
  <!---->
  <!---->
</div>
`;
